### DISCUSSION TITLE
如何配置才能启用全局ErrorHandler？

### DISCUSSION THREAD

**Discussion by sunjc** [tusen-ai/naive-ui#1191] [Q&A] [answered]
如何配置才能使app.config.errorHandler生效，应结合什么组件显示错误消息？

**Comment by 07akioni** [comment, marked as answer]
把 useMessage 挂在全局

**Reply by sunjc** [reply]
官方文档没有提供绑定app.config.globalProperties.$message的方法，而是建议挂载到window：
<script>
  import { useMessage } from 'naive-ui'
  import { defineComponent } from 'vue'

  // content
  export default defineComponent({
    setup() {
      window.$message = useMessage()
    }
  })
</script>

**Reply by 07akioni** [reply]
如果你想挂在全局也可以，自己做就好，总之注意 APP 渲染之前是不能渲染 message 的。

**Comment by sunjc** [comment]
Button组件没有使用emit定义事件，而是使用了props，最终通过下面的call方法调用：
`function call<A extends any[]> (funcs: Function[] | Function, ...args: A): void {
  if (Array.isArray(funcs)) {
    funcs.forEach((func) => (call as any)(func, ...args))
  } else return funcs(...args)
}`
这种方式会吞噬async/await的异常，不能被app.config.errorHandler捕获，希望尽快修改这一问题

**Reply by 07akioni** [reply]
从原理上看，async 和 await 的 error 就无法被 errorHandler 捕获。你可以弄一个 demo 我看看 emit 怎么捕获移步错误的。

**Reply by sunjc** [reply]
您可以简单的使用原生button，绑定click事件测试，也可以看看我的demo project：[heroes-vue](https://github.com/sunjc/heroes-vue)。不过这个demo需要启动后台才能运行，可选后台[heroes-deno](https://github.com/sunjc/heroes-deno)、[heroes-kotlin](https://github.com/sunjc/heroes-kotlin)或[heroes-api(Java)](https://github.com/sunjc/heroes-api)

**Reply by 07akioni** [reply]
https://codesandbox.io/s/practical-clarke-tthqh?file=/src/App.vue

什么都抓不到

**Reply by sunjc** [reply]
我修改了一下，您抽空再试试

**Reply by 07akioni** [reply]
给我一个 codesandbox 的例子

**Reply by sunjc** [reply]
在您的代码基础上改的，昨天忘发地址了 https://codesandbox.io/s/dlxc4?file=/src/App.vue

**Reply by 07akioni** [reply]
我研究了一下，emit 和 onXxx 的实现确实略有差别，这个我得研究一下有没有类似的方案

**Reply by 07akioni** [reply]
onXxx 的好处是可以拿到返回值，emit 不行。

Vue 在 emit 中使用了 `callWithAsyncErrorHandling` 所以可以捕获到这个错误。
